ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.cpp

```diff
@@ -10,37 +10,117 @@
+            Tensor b = ttnn::to_layout(a, Layout::TILE, std::nullopt, std::nullopt, (Device *)nullptr);
+            // pad c if needed
+            if (pad_c) {
+                ttnn::Shape padded_shape = b.get_shape().with_tile_padding();
+                ttnn::Shape shape = b.get_shape();
+                uint32_t C_rounded = round_up(padded_shape[1], 32);
+                b = ttnn::pad(b, std::array<uint32_t, 4>({padded_shape[0], C_rounded, padded_shape[2], padded_shape[3]}),
+                            std::array<uint32_t, 4>({0, 0, 0, 0}), 0);
+                b = ttnn::reshape(b, ttnn::Shape({shape[0], shape[1], shape[2], shape[3]}, {padded_shape[0], C_rounded, padded_shape[2], padded_shape[3]}));
+            }
+            // run the transpose
+            b = operation::run(Transpose{transpose_dim, output_mem_config}, {b}).at(0);
+            // slice back to original shape
+            if (b.get_shape()[1] != b.get_shape().with_tile_padding()[1] || b.get_shape()[0] != b.get_shape().with_tile_padding()[0]) {
+                std::array<uint32_t, 4> begins = {0, 0, 0, 0};
+                std::array<uint32_t, 4> ends = {b.get_shape()[0], b.get_shape()[1], b.get_shape().with_tile_padding()[2], b.get_shape().with_tile_padding()[3]};
+                std::array<uint32_t, 4> step = {1, 1, 1, 1};
+                return ttnn::slice(b, begins, ends, step);
+            }
+            // back to original layout
+            b = ttnn::to_layout(b, a.get_layout(), std::nullopt, std::nullopt, (Device *)nullptr);
+            return b;
+        }
         return operation::run(Transpose{transpose_dim, output_mem_config}, {a}).at(0);
     } else {
-        // TODO: Add pad_n to run_with_autoformat when needed
-        return operation::run_with_autoformat(Transpose{transpose_dim, output_mem_config}, {a}, {}, {}, 0, pad_c /*, pad_n */).at(0);
+        if (TransposeOpDim::HC == transpose_dim) {
+            Tensor b = a;
+            if (pad_c) {
+                tt::tt_metal::LegacyShape padded_shape = a.get_legacy_shape();
+                ttnn::Shape shape = a.get_shape();
+                uint32_t C_rounded = round_up(padded_shape[1], 32);
+                b = ttnn::pad(a, std::array<uint32_t, 4>({padded_shape[0], C_rounded, padded_shape[2], padded_shape[3]}),
+                            std::array<uint32_t, 4>({0, 0, 0, 0}), 0);
+                b = ttnn::reshape(b, ttnn::Shape({shape[0], shape[1], shape[2], shape[3]}, {padded_shape[0], C_rounded, padded_shape[2], padded_shape[3]}));
+            }
+
+            b = operation::run(Transpose{transpose_dim, output_mem_config}, {b}).at(0);
+
+            if (b.get_shape()[1] != b.get_shape().with_tile_padding()[1] || b.get_shape()[0] != b.get_shape().with_tile_padding()[0]) {
+                std::array<uint32_t, 4> begins = {0, 0, 0, 0};
+                std::array<uint32_t, 4> ends = {b.get_shape()[0], b.get_shape()[1], b.get_legacy_shape()[2], b.get_legacy_shape()[3]};
```

Comment:
`b.get_shape().with_tile_padding()` is equivalent to `b.get_legacy_shape()`, but please use `b.get_logical_shape()` and `b.get_padded_shape()` to avoid confusion